exchanges/bitmex/bitmex_websocket.go

```diff
@@ -131,338 +152,321 @@ func (b *Bitmex) wsReadData() {
 	if err != nil {
-		return err
+		return fmt.Errorf("unknown message format: %s", respRaw)
+	}
+	// We don't need to know about errors, since we're looking optimistically into the json
+	op, _ := jsonparser.GetString(msg, "request", "op")
+	errMsg, _ := jsonparser.GetString(msg, "error")
+	success, _ := jsonparser.GetBoolean(msg, "success")
+	version, _ := jsonparser.GetString(msg, "version")
+	switch {
+	case version != "":
+		op = "open"
+		fallthrough
+	case errMsg != "", success:
+		streamID, e2 := jsonparser.GetString(respRaw, "[1]")
+		if e2 != nil {
+			return fmt.Errorf("%w parsing stream", e2)
+		}
+		if !b.Websocket.Match.IncomingWithData(op+":"+streamID, msg) {
+			return fmt.Errorf("%w: %s:%s", stream.ErrNoMessageListener, op, streamID)
+		}
+		return nil
 	}
 
-	var respError WebsocketErrorResponse
-	if _, ok := quickCapture["status"]; ok {
-		err = json.Unmarshal(respRaw, &respError)
+	tableName, err := jsonparser.GetString(msg, "table")
+	if err != nil {
+		// Anything that's not a table isn't expected
+		return fmt.Errorf("unknown message format: %s", msg)
+	}
+
+	switch tableName {
+	case bitmexWSOrderbookL2, bitmexWSOrderbookL225, bitmexWSOrderbookL10:
+		var orderbooks OrderBookData
+		err = json.Unmarshal(msg, &orderbooks)
 		if err != nil {
 			return err
 		}
-	}
+		if len(orderbooks.Data) == 0 {
+			return fmt.Errorf("empty orderbook data received: %s", msg)
+		}
 
-	if _, ok := quickCapture["success"]; ok {
-		var decodedResp WebsocketSubscribeResp
-		err = json.Unmarshal(respRaw, &decodedResp)
+		var pair currency.Pair
+		var a asset.Item
+		pair, a, err = b.GetPairAndAssetTypeRequestFormatted(orderbooks.Data[0].Symbol)
 		if err != nil {
 			return err
 		}
 
-		if decodedResp.Success {
-			if len(quickCapture) == 3 {
-				if b.Verbose {
-					log.Debugf(log.ExchangeSys, "%s websocket: Successfully subscribed to %s",
-						b.Name, decodedResp.Subscribe)
-				}
-			} else {
-				b.Websocket.SetCanUseAuthenticatedEndpoints(true)
-				if b.Verbose {
-					log.Debugf(log.ExchangeSys, "%s websocket: Successfully authenticated websocket connection",
-						b.Name)
-				}
-			}
+		err = b.processOrderbook(orderbooks.Data, orderbooks.Action, pair, a)
+		if err != nil {
+			return err
+		}
+	case bitmexWSTrade:
+		if !b.IsSaveTradeDataEnabled() {
 			return nil
 		}
-
-		b.Websocket.DataHandler <- fmt.Errorf("%s websocket error: Unable to subscribe %s",
-			b.Name, decodedResp.Subscribe)
-	} else if _, ok := quickCapture["table"]; ok {
-		var decodedResp WebsocketMainResponse
-		err = json.Unmarshal(respRaw, &decodedResp)
+		var tradeHolder TradeData
+		err = json.Unmarshal(msg, &tradeHolder)
 		if err != nil {
 			return err
 		}
-		switch decodedResp.Table {
-		case bitmexWSOrderbookL2, bitmexWSOrderbookL225, bitmexWSOrderbookL10:
-			var orderbooks OrderBookData
-			err = json.Unmarshal(respRaw, &orderbooks)
+		var trades []trade.Data
+		for i := range tradeHolder.Data {
+			if tradeHolder.Data[i].Price == 0 {
```

Comment:
```suggestion
			if tradeHolder.Data[i].Size == 0 {
```

Reply:
Just going off from the commentary below

Reply:
Good catch 👍 
OutOfScope: I'll handle this separately 😄 
I'm going to be a bit stricter on not contaminating unrelated fixes, especially not into features.

Reply:
Fixed #1685